efi/preinstall/check_tpm.go

```diff
@@ -67,6 +68,54 @@ func openAndCheckTPM2Device(env internal_efi.HostEnvironment, flags checkTPM2Dev
 		savedTpm.Close()
 	}()
 
+	// Make sure that the TPM is not in failure mode, else the following GetCapability calls
+	// will fail due to the wrong property being returned (TPM2_GetCapability is usable in
+	// failure mode, but only to fetch information about the manufacturer, firmware version
+	// and vendor information, and it may round up the supplied property to match one of these).
+	// In addition to using TPM2_GetCapability to fetch this limited set of properties, only
+	// the TPM2_GetTestResult command is usable when the TPM is in failure mode.
+	// Trigger a self test of untested functionality (note that the platform firmware might
+	// have already done this, which is good because it will mean there's nothing to test).
+	err = tpm.SelfTest(false)
+	switch {
+	case tpm2.IsTPMWarning(err, tpm2.WarningTesting, tpm2.CommandSelfTest):
+		// This is an implementation that performs the remaining self tests after the
+		// command completes. We need to wait around to get a test result.
+		tick := time.NewTicker(100 * time.Millisecond)
+		for {
+			_, rc, err := tpm.GetTestResult()
+			if err != nil {
+				return nil, false, fmt.Errorf("cannot obtain self test results: %w", err)
+			}
+			if rc == tpm2.ResponseSuccess {
+				// All executed tests completed successfully.
+				break
+			}
+			switch rc {
+			case tpm2.ResponseFailure:
+				// One or more tests failed and the TPM is now in failure mode.
+				return nil, false, ErrTPMFailure
+			case tpm2.ResponseTesting:
+				// The tests are still running. We need to wait for a bit and
+				// then try again.
+			default:
+				return nil, false, fmt.Errorf("unexpected self test result: %#x", rc)
+			}
```

Comment:
do we need to timeout at some point and consider that a failure?

Reply:
We could do - I'm not sure what to set it to though.

Reply:
maybe it needs to be passed in from outside? do we have a sense of how long this kind of self-testing takes?

Reply:
I don't - there are many factors here - eg, firmware based (faster) vs discrete (likely slower) and what tests are left to run (did the firmware execute `TPM2_SelfTest` before, or has functionality been self-tested incrementally as it's used?).

For now, I've commented out this particular code path with a comment explaining why it's likely to be ok to not have it. I've left the code there though for now, and we can uncomment it and re-open the timeout question another day if we discover that we do actually need it - I suspect the best thing will be to just use the `context.Context` that's already passed into the public API.